presto-common/src/main/java/com/facebook/presto/common/block/ColumnarMap.java

```diff
@@ -0,0 +1,240 @@
+                //     the average row size: keysBlock.getSizeInBytes() / (double) keysBlock.getPositionCount() + valuesBlock.getSizeInBytes() / (double) valuesBlock.getPositionCount() * the number of rows: offsets[positionCount]
+                (Integer.BYTES + Byte.BYTES) * positionCount +
+                        (long) ((keysBlock.getPositionCount() == 0 ? 0 : keysBlock.getSizeInBytes() / (double) keysBlock.getPositionCount() + valuesBlock.getPositionCount() == 0 ? 0 : valuesBlock.getSizeInBytes() / (double) valuesBlock.getPositionCount()) * offsets[positionCount]));
+    }
+
+    private static ColumnarMap toColumnarMap(RunLengthEncodedBlock rleBlock)
+    {
+        ColumnarMap columnarMap = toColumnarMap(rleBlock.getValue());
+        int positionCount = rleBlock.getPositionCount();
+
+        // build new offsets block
+        int[] offsets = new int[positionCount + 1];
+        int entryCount = columnarMap.getEntryCount(0);
+        for (int i = 0; i < offsets.length; i++) {
+            offsets[i] = i * entryCount;
+        }
+
+        // create indexes for a dictionary block of the elements
+        int[] dictionaryIds = new int[positionCount * entryCount];
+        int nextDictionaryIndex = 0;
+        for (int position = 0; position < positionCount; position++) {
+            for (int entryIndex = 0; entryIndex < entryCount; entryIndex++) {
+                dictionaryIds[nextDictionaryIndex] = entryIndex;
+                nextDictionaryIndex++;
+            }
+        }
+
+        Block keysBlock = columnarMap.getKeysBlock();
+        Block valuesBlock = columnarMap.getValuesBlock();
+        return new ColumnarMap(
+                rleBlock,
+                0,
+                offsets,
+                new DictionaryBlock(dictionaryIds.length, keysBlock, dictionaryIds),
+                new DictionaryBlock(dictionaryIds.length, valuesBlock, dictionaryIds),
+                columnarMap.getHashTables(),
+                INSTANCE_SIZE + rleBlock.getRetainedSizeInBytes() + sizeOf(offsets) + sizeOf(dictionaryIds),
+                // The estimated serialized size is the sum of the following
+                // 1) the offsets size: Integer.BYTES * positionCount
+                // 2) nulls array size: Byte.BYTES * positionCount
+                // 3) the estimated serialized size for the keysBlock and valuesBlock which were just constructed as new DictionaryBlocks:
+                //     the average row size: keysBlock.getSizeInBytes() / (double) keysBlock.getPositionCount() + valuesBlock.getSizeInBytes() / (double) valuesBlock.getPositionCount() * the number of rows: offsets[positionCount]
+                (Integer.BYTES + Byte.BYTES) * positionCount + (long) ((keysBlock.getSizeInBytes() / (double) keysBlock.getPositionCount() + valuesBlock.getSizeInBytes() / (double) valuesBlock.getPositionCount()) * offsets[positionCount]));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).